pkg/cri/sbserver/container_create.go

```diff
@@ -375,3 +395,216 @@ func (c *criService) runtimeSnapshotter(ctx context.Context, ociRuntime criconfi
 	log.G(ctx).Debugf("Set snapshotter for runtime %s to %s", ociRuntime.Type, ociRuntime.Snapshotter)
 	return ociRuntime.Snapshotter
 }
+
+const (
+	// relativeRootfsPath is the rootfs path relative to bundle path.
+	relativeRootfsPath = "rootfs"
+	// hostnameEnv is the key for HOSTNAME env.
+	hostnameEnv = "HOSTNAME"
+)
+
+// buildContainerSpec build container's OCI spec depending on controller's target platform OS.
+func (c *criService) buildContainerSpec(
+	platform *types.Platform,
+	id string,
+	sandboxID string,
+	sandboxPid uint32,
+	netNSPath string,
+	containerName string,
+	imageName string,
+	config *runtime.ContainerConfig,
+	sandboxConfig *runtime.PodSandboxConfig,
+	imageConfig *imagespec.ImageConfig,
+	extraMounts []*runtime.Mount,
+	ociRuntime config.Runtime,
+) (_ *runtimespec.Spec, retErr error) {
+	var (
+		specOpts []oci.SpecOpts
+
+		// Platform helpers
+		isLinux   = platform.OS == "linux"
+		isWindows = platform.OS == "windows"
+	)
+
+	if isLinux {
```

Comment:
What about using the code style as you mentioned in the description? like

```go
func buildContainerSpec(...) {
     case isLinux:
          return buildLinuxContainerSpec()
     case isWindows:
          return buildWindowsContainerSpec()
     default:
            return error
     }
}
```

The `container_create_windows.go` can be replaced by `windows_container_create.go` so that the spec builder can run in any platform.

But in this commit, the multiple platforms codepath are mixed. It is hard to review and maintain.

Reply:
+1, the constant flip between isLinux and isWindows is jarring. Although perhaps just return the spec opts that only apply to that platform? Otherwise their may be a lot of copy+paste. e.g. linux can return a slice with oci.WithoutRunMount, hostname env var set etc.

Reply:
Linux spec is quite complex and order is important in many places, so it's either many `if` branches or separate functions. I'm ok with having a separate builder for each platform in favor of readability.